### DISCUSSION TITLE
bug: usePrepareContractWrite cache should clear if another transaction has succeeded

### DISCUSSION THREAD

**Discussion by KholdStare** [wevm/wagmi#2389] [Ideas]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

0.12.8

### Current Behavior

Currently if there are two separate transactions, where one can depend on the other, `usePrepareContractWrite` result is cached for both transactions. When one of the transaction completes, the other needs to be refetched manually, otherwise the cached result stays, even if the actual result would now differ.

An example of this is if you have an ERC20 approval transaction and a use of it in two different components.

Approval:

```
 const { config } = usePrepareContractWrite({
    address: token.data?.address,
    abi: erc20ABI,
    functionName: "approve",
    args: [approveAddress, constants.MaxUint256],
  });

  const { writeAsync: writeAllowance } = useContractWrite(config);
```

Buy something with ERC20:

```
 const prepareResult = usePrepareContractWrite({
    address,
    enabled,
    abi,
    functionName: "buy",
    args,
  });
```

The problem is, there isn't anything there to trigger `usePrepareContractWrite` to re-evaluate after the approval is done. This has to be somehow manually connected, e.g. only enabling it if the approval has succeeded.

My thinking is that the WAGMI context should invalidate the caches of all `usePrepareContractWrite` calls when any other transaction succeeds - otherwise they may be assuming a stale state of the blockchain.

### Expected Behavior

Caches of all `usePrepareContractWrite` calls should be invalidated when any other transaction succeeds.

**Comment by tmm** [comment]
Converted to feature request. You can try changing the [`cacheTime`](https://wagmi.sh/react/prepare-hooks/usePrepareContractWrite#cachetime-optional) and [`staleTime`](https://wagmi.sh/react/prepare-hooks/usePrepareContractWrite#staletime-optional).

**Reply by KholdStare** [reply]
I feel like a `watch` or `cacheOnBlock` type of functionality, but on `usePrepareContractWrite` would be great. Currently those are only available on `useContractRead`.

We got "around" the problem by triggering the `refetch` on the `usePrepareContractWrite` result, when `allowance` changed in a different `useContractRead` result that had the `watch` option enabled.
